typed/merge.go

```diff
@@ -17,11 +17,62 @@ limitations under the License.
 package typed
 
 import (
+	"sort"
+
 	"sigs.k8s.io/structured-merge-diff/fieldpath"
 	"sigs.k8s.io/structured-merge-diff/schema"
 	"sigs.k8s.io/structured-merge-diff/value"
 )
 
+type pathElementValueMap struct {
```

Comment:
When I wrote the set version of this, getting the sort call and less function to do the right thing was tricky. I think unit tests of this particular thing are in order to keep it working :/

Alternatively consider moving this into the same package as the set type where it'd be easier to share code and tests?

Reply:
Yeah, we're doing very similar things everywhere, but I'd also like to avoid having `PathElementMap` to `interface{}` (which we could use in all three places)...

Reply:
Why? I think we need either tests or shared implementation. I don't think interface{} is going to make it slow if that's what you're worried about?

Reply:
Yeah, I suspect it has to escape to heap when you cast to `interface{}`, but I might be wrong, it's worth trying.